### DISCUSSION TITLE
Best parameters settings - range - data rate - reliability

### DISCUSSION THREAD

**Discussion by ImTandoff** [sandeepmistry/arduino-LoRa#608] [Q&A] [answered]
Hello everyone, I am currently working in a project where sensores and motors were connected trough a RS485 communication. The slaves were far from the Master, and the aplicattion were agriculture related, so , LoRa tecnology was a card on the table. 

The module used is the RFM995 with a ESP32. 

The first step was to check the maximum range of the tecnlogy. So, once the connection was set, a distance of 300m with the default settings wich is good good but nothing comperate with some other test were people could achieve 5km. In my situation it was a non obstacle scenario, what makes me think that I should achieve more distance. Trying with the following settings 450m were achieve but with a poor RSSI indicator.

//-----MODULATION PARAMETERS----->
#define TX_POWER 20
#define SF 9
#define BANDWIDTH 125E3
#define SYNCWORD 0xF3
#define CODERATE 1
#define PREAMBLELENGTH 

Does anyone know what i could do to reach at least 1 km with an RSSI of -70?

**Comment by Kongduino** [comment, marked as answer]
It depends on so many variables. But here are a few indications:

- Good antennas on both side.
- The higher the SF, the lower the BW, the better.
- You could also try playing with C/R – from 4/5 to 4/8 for example.
- PA_BOOST, LNA_Gain

I wrote an app that writes setup code for this library. Here's the end product, after some tweaking:

```c
// BW = 6: 62.5 kHz, CR = 1: 4/5, HM = 0
uint8_t reg1 = 0x62;
// SF = 12: 12, CRC = 0
uint8_t reg2 = 0xC0;
// LDRO = 1, AGCAutoOn = 1
uint8_t reg3 = 0x0C;
// PaSelect = 1, MaxPower = 7: 15 dBm, OutputPower = 15: 17 dBm
uint8_t regpaconfig = 0xFF;
#define REG_OCP 0x0B
#define REG_PA_CONFIG 0x09
#define REG_LNA 0x0c
#define REG_OP_MODE 0x01
#define REG_MODEM_CONFIG_1 0x1d
#define REG_MODEM_CONFIG_2 0x1e
#define REG_MODEM_CONFIG_3 0x26
#define REG_PA_DAC 0x4D
#define PA_DAC_HIGH 0x87
#define MODE_LONG_RANGE_MODE 0x80
#define MODE_SLEEP 0x00
#define MODE_STDBY 0x01
#define MODE_TX 0x03
#define MODE_RX_CONTINUOUS 0x05
#define MODE_RX_SINGLE 0x06
LoRa.writeRegister(REG_OP_MODE, MODE_LONG_RANGE_MODE | MODE_SLEEP);
delay(10);
LoRa.writeRegister(REG_PA_CONFIG, regpaconfig);
LoRa.writeRegister(REG_MODEM_CONFIG_1, reg1);
LoRa.writeRegister(REG_MODEM_CONFIG_2, reg2);
LoRa.writeRegister(REG_MODEM_CONFIG_3, reg3);
delay(10);
LoRa.writeRegister(REG_OP_MODE, MODE_LONG_RANGE_MODE | MODE_STDBY);
```

However it requires a small modification of the library. The two lines [there](https://github.com/sandeepmistry/arduino-LoRa/blob/master/src/LoRa.h#L109-L110) should be moved out of `private` and set to public.

```c
  uint8_t readRegister(uint8_t address);
  void writeRegister(uint8_t address, uint8_t value);
```

Move them to line94 or so, and this will work.

`-70 RSSI` is a little too hopeful but you should get good results, and good SNR too, which is just as important.

**Reply by ImTandoff** [reply]
Thanks @Kongduino. Do you know where I could get more information about PA_BOOST, LNA_Gain?

**Reply by Kongduino** [reply]
The first place to look at is, always, [the datasheet](https://semtech.my.salesforce.com/sfc/p/#E0000000JelG/a/2R0000001Rbr/6EfVZUorrpoKFfvaF_Fkpgp5kzjiNyiAbqcpqh9qSjE) [PDF]. After that you can google around, but where the sx1276 is concerned this document is your bible. :-)
